src/Parser.ts

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) ZALORA
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+import Lexer from './Lexer';
+import ASTExpression from './AST/ASTExpression';
+import { lbracket, dot, ident, int, rbracket, eof, Token } from './tokens';
+import ASTStatement from './AST/ASTStatement';
+import ASTExpressionStatement from './AST/ASTExpressionStatement';
+import ASTSelectExpression from './AST/ASTSelectExpression';
+import ASTIndexExpression from './AST/ASTIndexExpression';
+import ASTIntegerLiteral from './AST/ASTIntegerLiteral';
+import ASTIdentifier from './AST/ASTIdentifier';
+import Program from './AST/Program';
+
+enum Precedence {
+  lowest,
+  index,
+}
+
+type PrefixParser = () => ASTExpression | undefined;
+
+type InfixParser = (arg?: ASTExpression) => ASTExpression | undefined;
+
+class ParserError {
+  private message: string;
+
+  constructor(message: string) {
+    this.message = message;
+  }
+
+  getMessage(): string {
+    return this.message;
+  }
+}
+
+class JSLParserError extends Error {
+  private errors: Array<ParserError> = [];
+
+  constructor(errors: Array<ParserError>) {
+    super();
+    this.errors = errors;
+
+    // Set the prototype explicitly.
+    Object.setPrototypeOf(this, JSLParserError.prototype);
```

Comment:
The prototype should be set explicitly so that the methods declared inside `JSLParserError ` can be accessible. This happens when we are subclassing the user defined class with `Error` and others, reference [here](https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work).